src/main/java/net/neoforged/neoforge/items/ItemStackHandler.java

```diff
@@ -5,18 +5,41 @@
 
 package net.neoforged.neoforge.items;
 
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.HolderLookup;
 import net.minecraft.core.NonNullList;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.nbt.Tag;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
+import net.neoforged.neoforge.codec.NonNullListCodecs;
 import net.neoforged.neoforge.common.util.INBTSerializable;
 
 public class ItemStackHandler implements IItemHandler, IItemHandlerModifiable, INBTSerializable<CompoundTag> {
     protected NonNullList<ItemStack> stacks;
 
+    public static final Codec<ItemStackHandler> CODEC = RecordCodecBuilder.create(i -> i.group(
+            NonNullListCodecs.withIndices(ItemStack.OPTIONAL_CODEC, ItemStack.EMPTY, "slot", ItemStack::isEmpty)
+                    .fieldOf("stacks").forGetter(handler -> handler.stacks))
+            .apply(i, ItemStackHandler::new));
+
+    /**
+     * Use if you require 1-to-1 support for the output of {@link #serializeNBT(HolderLookup.Provider)}.
+     * Otherwise, prefer {#link CODEC} as a newer alternative to saving data to disk.
+     * 
+     * @param provider
+     * @return
+     */
+    @Deprecated
+    public static Codec<ItemStackHandler> nbtCompatibleCodec(HolderLookup.Provider provider) {
```

Comment:
Is the only intended usecase for this to be serialized to NBT? If it's serialized to JSON, we may want to put some safeguards in place to prevent crashes during number list conversions.

Reply:
It may be useful to convert old NBT data into JSON for backups or something. Why guard it?

Reply:
The issue occurs when converting JSON number lists to NBT. Basically, the conversion that `NBTOps` uses checks for the lowest number of bytes needed to represent a number. So, if your first number starts with 8, the ops will create a byte list tag. However, if your numbers are not all represented by bytes in this example, it will crash. For example, converting the list [8, -1, 32914] will crash. So, there needs to be an extra safeguard when transforming between the two.